src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -141,146 +272,138 @@ public async Task UpdatePrimaryBranchSolutionAsync(AssetProvider assetProvider,
+                solution = solution.AddProject(projectInfo);
+            return solution;
         }
 
         /// <summary>
-        /// Adds an entire solution to the workspace, replacing any existing solution.
+        /// Attempts to update this workspace with the given <paramref name="newSolution"/>.  If this succeeds, <see
+        /// langword="true"/> will be returned in the tuple result as well as the actual solution that the workspace is
+        /// updated to point at.  If we cannot update this workspace, then <see langword="false"/> will be returned,
+        /// along with the solution passed in.
         /// </summary>
-        internal bool TrySetCurrentSolution(SolutionInfo solutionInfo, int workspaceVersion, SerializableOptionSet options, [NotNullWhen(true)] out Solution? solution)
+        private async ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceAsync(
+            int workspaceVersion,
+            bool fromPrimaryBranch,
+            Solution newSolution,
+            CancellationToken cancellationToken)
         {
-            lock (_currentSolutionGate)
+            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
             {
-                if (workspaceVersion <= _currentRemoteWorkspaceVersion)
-                {
-                    // we never move workspace backward
-                    solution = null;
-                    return false;
-                }
+                var oldSolution = this.CurrentSolution;
+
+                // if this wasn't from the primary branch, then we have nothing to do.  Just return the solution back for
+                // the caller.
+                //
+                // we never move workspace backward
+                if (!fromPrimaryBranch || workspaceVersion <= _currentRemoteWorkspaceVersion)
+                    return (newSolution, updated: false);
 
-                // set initial solution version
                 _currentRemoteWorkspaceVersion = workspaceVersion;
 
-                // clear previous solution data if there is one
-                // it is required by OnSolutionAdded
-                ClearSolutionData();
+                // if either solution id or file path changed, then we consider it as new solution. Otherwise,
+                // update the current solution in place.
 
-                OnSolutionAdded(solutionInfo);
+                var addingSolution = oldSolution.Id != newSolution.Id || oldSolution.FilePath != newSolution.FilePath;
+                if (addingSolution)
+                {
+                    // We're not doing an update, we're moving to a new solution entirely.  Clear out the old one. This
+                    // is necessary so that we clear out any open document information this workspace is tracking. Note:
+                    // this seems suspect as the remote workspace should not be tracking any open document state.
+                    this.ClearSolutionData();
```

Comment:
@tmat for thoughts. t his preserves what we had before... but hwat we had before seems very fishy/suspect.

Reply:
Yeah, let's keep it for now and just remove the workspace updating completely once it's possible.

Reply:
👍